Implement serialization and deserialization for ISIN

korrat opened the pull request:
Fixes part of #1.

As noted in the code, there are two issues for discussion:

1. The current code uses `ISIN::value` for serialization, which is marked as deprecated. However, the suggested `ToString::to_string` is not a good replacement, since it necessitates a heap allocation. There are three options for resolution:
   1. Do nothing & keep as is. May run into trouble when removing `ISIN::value` at some point.
   2. Change serialization to construct the `str` slice on the fly, as done in other methods.
   3. Implement `AsRef<str>` for ISIN and advertise that as replacement for `ISIN::value`.
   4. Undeprecate `ISIN::value`
2. Right now, (de)serialization always uses strings. For formats that support it, we could serialize an ISIN as a fixed-length byte array, potentially saving some overhead.

gnp commented:
Thank you. Answering the questions... (1) `value` was deprecated because I wanted to use existing Rust practice where possible and the underlying value is the array not a string; `AsRef<str>` is an attractive improvement in line with that. (2) I expect {,de}serialization {to,from} strings to be the main use case; let's get that working separately.

Could you provide a program in the examples directory that uses the serde support and demonstrates json serialization and deserialization?

korrat commented:
I've added a basic example of the serialization. Let me know if you would like something more elaborate.

I've also added the `AsRef` implementation and serialization uses that now. If you prefer, I can split that out into a separate PR.

korrat commented:
I've kept individual commits for now, but if you prefer, I can squash some or all commits before the merge.

I'd appreciate it if you would consider marking this PR as `hacktoberfest-accepted` after merging so that it counts for Hacktoberfest.

gnp commented:
Happy to tag for hacktoberfest. Reading up on that now. I added "topic" tag "hacktoberfest" to the repo itself as it appears that might be required. Checking for how to mark the PR appropriately once accepted.

gnp commented on src/lib.rs:
```diff
@@ -308,8 +306,8 @@ impl serde::Serialize for ISIN {
     {
-        // TODO we could decide to serialize as bytes here, using `serializer.is_human_readable()`
-        serializer.serialize_str(self.value())
+        // TODO we could decide to serialize as byte array here, using `serializer.is_human_readable()`
```
Let's remove this comment. We can have an "issue" filed that proposes support for byte array serialization and we can discuss in that issue what it would look like to use that vs string serialization as a user of the crate.

gnp commented:
I made one final request in a comment on _src/lib.rs_. A separate PR for the `AsRef<str>` implementation is not necessary. Squashing would be appreciated.

korrat commented:
I removed the comment and squashed the commits. Next, I will add that issue about byte array serialization.

As for Hacktoberfest, adding the `hacktoberfest` topic to the repo is already enough to count the contribution. The `hacktoberfest-accepted` label is an alternative to opt in for a single PR.

gnp commented:
Fixes #1

gnp commented:
There are other identifier crates in the series and I make parallel improvements in them all so their DX remains the same.

I plan to make corresponding changes in the other crates.

If you want to get credit for those too I could mark those repos with hacktoberfest as well and you could make PRs with the equivalent changes.

Either way is fine with me. I appreciate you contribution! LMK

gnp commented:
Also @korrat I just cut the 0.1.17 release of isin crate including your changes.
